tests/integration/test_l1_write_request.py

```diff
@@ -0,0 +1,254 @@
+        self.assertEqual(self.cache.statistics.cache_misses, 1)
+
+    def test_modified(self):
+        """
+        Test writing a line in the Modified state.
+        Other caches return a NOHIT snoop response.
+
+        Requires the following sequence:
+        - Write line (cache miss)
+        - Write same line (cache hit)
+        """
+        # Event 1: Write miss at self.nohit_addr (gets line in M state)
+        handle_event(self.cache, self.trace_event, self.nohit_addr)
+
+        # Confirm the RWIM bus op was issued
+        self.assert_log_called_once_with(LogLevel.NORMAL, r"busop.*(rwim|4).*")
+
+        # Confirm the L2 sendline message was issued for L1 request
+        self.assert_log_called_once_with(LogLevel.NORMAL, r"l2.*sendline.*")
+
+        self.check_line_state(self.nohit_addr, MESIState.MODIFIED)
+
+        # Event 1: Write hit at self.nohit_addr (should stay in M state)
+        handle_event(self.cache, self.trace_event, self.nohit_addr)
+        self.check_line_state(self.nohit_addr, MESIState.MODIFIED)
+
+        # Confirm the L2 sendline message was issued for each L1 request
+        self.assert_log_called_with_count(LogLevel.NORMAL, r"l2.*sendline.*", 2)
+
+        # TODO: Decide whether or not we want to assert additional busops
+        # have *not* been issued, or if this one is even of value.
+        self.assert_log_called_with_count(
+            LogLevel.NORMAL, r"busop.*(write|2).*address.*", 0
+        )
+
+        # Assertions for statistics
+        self.assertEqual(self.cache.statistics.cache_reads, 0)
+        self.assertEqual(self.cache.statistics.cache_writes, 2)
+        self.assertEqual(self.cache.statistics.cache_hits, 1)
+        self.assertEqual(self.cache.statistics.cache_misses, 1)
+
+    def test_shared(self):
+        """
+        Test writing a line in the Shared state.
+        Other caches return a HIT snoop response.
+
+        Requires the following sequence:
+        - Read line (cache miss)
+        - Write line (cache hit)
+        """
+        # Event 0: Read miss at self.hit_addr (gets line in S state)
+        handle_event(self.cache, TraceCommand.L1_DATA_READ, self.hit_addr)
+        self.check_line_state(self.hit_addr, MESIState.SHARED)
+
+        # Event 1: Write hit at self.hit_addr (should move to M state)
+        handle_event(self.cache, self.trace_event, self.hit_addr)
+        self.check_line_state(self.hit_addr, MESIState.MODIFIED)
+
+        # Confirm the INVALIDATE bus op was issued
+        self.assert_log_called_with_count(
+            LogLevel.NORMAL, r"busop.*(invalidate|3).*", 1
+        )
+
+        # Confirm the L2 sendline message was issued for each L1 request
+        self.assert_log_called_with_count(LogLevel.NORMAL, r"l2.*sendline.*", 2)
+
+        # Only the first read request should have resulted in a READ bus operation.
+        self.assert_log_called_once_with(LogLevel.NORMAL, r".*busop.*(read|1).*address")
```

Comment:
Just FYI, I had to start getting more specific with this regex in order not to incorrectly detect other bus ops. We'll want to keep that in mind when creating the snoop request integration tests.